### PR TITLE
Pivotal #ID 172650484: Add support for folders in submission

### PR DESCRIPTION
**Pull request by Juan-EBI**
https://www.pivotaltracker.com/story/show/172650484

### CODE REVIEW

File: commons/commons-util/src/main/kotlin/ebi/ac/uk/io/FileUtils.kt [hunk 1]
```diff
@@ -30,6 +43,21 @@ object FileUtils {
         Files.write(createParentDirectories(source.toPath()), content.toByteArray())
     }
 
+    fun isDirectory(file: File): Boolean {
```
Comment (jhoanmanuelms) [inline commons/commons-util/src/main/kotlin/ebi/ac/uk/io/FileUtils.kt:46]: one-liner

File: commons/commons-util/src/main/kotlin/ebi/ac/uk/io/FileUtils.kt [hunk 2]
```diff
@@ -19,7 +20,19 @@ object FileUtils {
     }
 
     fun copyOrReplaceFile(source: File, target: File) {
-        Files.copy(source.toPath(), createParentDirectories(target.toPath()), StandardCopyOption.REPLACE_EXISTING)
+        when (source.isDirectory) {
+            true -> copyFolder(source.toPath(), target.toPath())
```
Comment (jhoanmanuelms) [inline commons/commons-util/src/main/kotlin/ebi/ac/uk/io/FileUtils.kt:24]: I prefer the way you used it below with true/else branches

File: commons/commons-util/src/main/kotlin/ebi/ac/uk/io/FileUtils.kt [hunk 3]
```diff
@@ -30,6 +43,21 @@ object FileUtils {
...
+    }
+
+    fun listFiles(file: File): List<File> {
+        return when (isDirectory(file)) {
+            true -> Files.list(file.toPath()).map { it.toFile() }.toList()
+            else -> emptyList()
+        }
+    }
+
+    fun size(file: File): Long {
```
Comment (jhoanmanuelms) [inline commons/commons-util/src/main/kotlin/ebi/ac/uk/io/FileUtils.kt:57]: one-liner

File: commons/commons-util/src/main/kotlin/ebi/ac/uk/io/ext/FileExtensions.kt [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+fun File.asFileList(): List<File> = FileUtils.listFiles(this)
+
+fun File.size() = FileUtils.size(this)
+
+fun File.createDirectory(name: String): File = Files.createDirectory(toPath().resolve(name)).toFile()
+
+fun File.createNewFile(name: String): File = resolve(name).apply { createNewFile() }
+
+fun File.createNewFile(name: String, text: String): File = resolve(name).apply { createNewFile(); writeText(text) }
```
Comment (jhoanmanuelms) [inline commons/commons-util/src/main/kotlin/ebi/ac/uk/io/ext/FileExtensions.kt:17]: Why don't leave each method call on its own line? A semi-colon looks very odd in Kotlin :sweat_smile:

File: commons/commons-util/src/test/kotlin/ebi/ac/uk/io/FileUtilsTest.kt [hunk 5]
```diff
@@ -17,6 +19,78 @@ internal class FileUtilsTest(private val temporaryFolder: TemporaryFolder) {
...
+
+        @Nested
+        inner class WhenFolder {
+
+            @Test
+            fun whenTargetFolderExists() {
+                val tempDir = temporaryFolder.createDirectory("directory")
+                tempDir.createNewFile("two.txt")
+
+                val targetDirectory = temporaryFolder.createDirectory("directory-target")
```
Comment (jhoanmanuelms) [inline commons/commons-util/src/test/kotlin/ebi/ac/uk/io/FileUtilsTest.kt:57]: typo: "target-directory"

File: submission/persistence/src/main/kotlin/ac/uk/ebi/biostd/persistence/model/DbFile.kt [hunk 6]
```diff
@@ -31,6 +31,8 @@ class DbFile(
 
     var size: Long = 0L
 
+    var directory: Boolean = false
```
Comment (jhoanmanuelms) [inline submission/persistence/src/main/kotlin/ac/uk/ebi/biostd/persistence/model/DbFile.kt:34]: You should update the database schema (Schema.sql) with this change. Also, shouldn't we make these changes for the ReferencedFile object, which is used by the files lists?

File: commons/commons-util/src/main/kotlin/ebi/ac/uk/io/CopyFileVisitor.kt [hunk 7]
```diff
@@ -0,0 +1,19 @@
+package ebi.ac.uk.io
+
+import java.nio.file.FileVisitResult
+import java.nio.file.Files
+import java.nio.file.Path
+import java.nio.file.SimpleFileVisitor
+import java.nio.file.attribute.BasicFileAttributes
+
+internal class CopyFileVisitor(private var sourcePath: Path, private val targetPath: Path) : SimpleFileVisitor<Path>() {
```
Comment (jhoanmanuelms) [inline commons/commons-util/src/main/kotlin/ebi/ac/uk/io/CopyFileVisitor.kt:9]: Nice solution :+1:

**Comment by jhoanmanuelms** [comment]
Please remember changing the destination branch to `dev`

File: submission/persistence/src/main/kotlin/ac/uk/ebi/biostd/persistence/model/DbFile.kt (hunk 6 above)
Comment (Juan-EBI) [reply submission/persistence/src/main/kotlin/ac/uk/ebi/biostd/persistence/model/DbFile.kt:34]: Actually this was supported so field is already there. I will create a ticket to add support in file list

Comment (jhoanmanuelms) [reply submission/persistence/src/main/kotlin/ac/uk/ebi/biostd/persistence/model/DbFile.kt:34]: It's in the database but it's not in the Schema.sql file

https://github.com/EBIBioStudies/biostudies-backend-services/blob/master/infrastructure/src/main/resources/setup/database/Schema.sql#L28

**Pull request merged**